src/parsers/gene_association_file.py

```diff
@@ -0,0 +1,327 @@
+        print(df.loc[invalid_mask, "annotation_date"].unique())
+
+    # Convert valid parts to standard date format
+    df.loc[~invalid_mask, "annotation_date"] = pd.to_datetime(
+        date_strs[~invalid_mask], format="%Y%m%d", errors="coerce").dt.strftime("%Y-%m-%d")
+    df.loc[invalid_mask, "annotation_date"] = None
+    return df
+
+def process_predicates(df):
+    """
+    Detect negative relationships in 'predicate' (starting with 'NOT|')
+    Clean the 'predicate' field into canonical GO relationship names
+    """
+    predicates = df["predicate"].astype(str)
+    df["negated"] = predicates.str.startswith("NOT|")
+    df["predicate"] = predicates.str.replace(r"^NOT\|", "", regex=True)
+    return df
+
+def transform_go_data(df):
+    """
+    Transform a raw GO annotation DataFrame into a standardized format for downstream use.
+    - Renames specific GAF columns to semantic schema-compliant names.
+    - Normalizes annotation dates.
+    - Adds fixed metadata columns ('aggregator', 'protocol_id').
+    - Splits pipe-delimited strings in 'publications' and 'supporting_objects' into lists.
+    - Processes and validates predicates using domain rules.
+    - Constructs the 'subject' field by combining 'DB' and 'DB_Object_ID'.
+    - Ensures all required columns are present (inserting None where missing).
+    - Reorders the columns to match the expected output format.
+    """
+    df.rename(columns={
+        "Qualifier": "predicate",
+        "GO_ID": "object",
+        "DB_Reference": "publications",
+        "With_From": "supporting_objects",
+        "Date": "annotation_date",
+        "Assigned_By": "primary_knowledge_source"
+    }, inplace=True)
+
+    df = normalize_dates(df)
+    df["aggregator"] = "UniProt"
+    df["protocol_id"] = None
+    
+    # Split '|' separated publication/supporting_objects into lists
+    df["publications"] = df["publications"].apply(lambda x: str(x).split('|') if pd.notna(x) and isinstance(x, str) else x)
+    df["supporting_objects"] = df["supporting_objects"].apply(lambda x: str(x).split('|') if pd.notna(x) and isinstance(x, str) else x)
+    df = process_predicates(df)
+
+    # Validate predicates
+    invalid_predicates = ~df["predicate"].isin(ALLOWED_PREDICATES)
+    if invalid_predicates.any():
+        raise ValueError(f"Invalid predicates found: {df.loc[invalid_predicates, 'predicate'].unique()}")
+    
+    df["subject"] = df["DB"].astype(str) + ":" + df["DB_Object_ID"].astype(str)
+
+    DESIRED_COLUMN_ORDER = [
+        "subject", "DB", "DB_Object_ID", "predicate", "object", 
+        "publications", "Evidence_Code","supporting_objects", 
+        "annotation_date", "primary_knowledge_source","aggregator", 
+        "protocol_id", "negated"]
+    
+    for col in DESIRED_COLUMN_ORDER:
+        if col not in df.columns:
+            df[col] = None #fill missing column with None
+
+    # Reorder columns to match desired output format
+    df = df[DESIRED_COLUMN_ORDER]
```

Comment:
the columns don't need to be in any specific order if the output file has a header line, so this line is not really needed